doc/dev/adr/0020-error-modeling-headers.md

```diff
@@ -0,0 +1,103 @@
+# ADR 19: Modeling User Errors
+
+## Context
+
+We have received feedback from users that they would like to have RPC responses communicate application level failures in the headers of the response rather than the payload of the response. 
+
+This feedback is, in part, because some applications want to route RPC responses without reading/deserializing the entire payload. It is also partly because [some applications cannot extend or change the payload model to accommodate error handling.](https://github.com/Azure/iot-operations-sdks/issues/488#issuecomment-2707496996).
+
+## Decision
+
+Our SDKs will add APIs on the command executor side that mark an RPC response as an "application error" by attaching two new MQTT user properties ("AppErrCode" and "AppErrPayload") whose values are a user-defined string and a user-defined payload object (serialized to bytes and then encoded as a UTF8 string) respectively. When a user wants to indicate that an RPC call failed with an application error they must provide the "AppErrCode" value, but "AppErrPayload" is optional.
+
+In addition, users must be able to include arbitrary user properties in their command response to support this error reporting in case our standard fields are insufficient. This feature is likely in place already for all languages.
+
+On the command invoker side, we will add APIs for checking if a response was an application error and returning the error code and error data fields parsed from the MQTT message "AppErrCode" and "AppErrPayload" user properties.
+
+Similar to how our SDKs handle serializing the actual MQTT message payload, our SDKs will require the user provide the serializer for serializing/deserializing the AppErrPayload object.
+
+Other than these two new user properties, the over-the-wire behavior of our protocol won't change as a result of this decision.
+
+In order to provide a strongly-typed experience, we will also add codegen support for modeling both the possible error codes (enum-like list of possible values?) and the type of the error payload in DTDL. This modeling will be detailed in a separate ADR, though.
+
+By convention, the value of the AppErrPayload object should be a UTF-8 encoded JSON string since MQTT user property values must be UTF-8 encoded strings.
```

Comment:
Above is state that the Error object is serialized to bytes and encoded as UTF8 string, this is mismatching with the UTF-8 encoded JSON string

Reply:
Good catch. I'll rewrite this.